app/src/test/java/de/westnordost/streetcomplete/quests/opening_hours/OpeningHoursTagParserTest.kt

```diff
@@ -0,0 +1,174 @@
+        assertNull(OpeningHoursTagParser.parse("Mo[1] 09:00-18:30")) // first monday within month
+        assertNull(OpeningHoursTagParser.parse("Mo 06:00+")) // open ended, without end hour at all
+        assertNull(OpeningHoursTagParser.parse("week 01-51 Mo 06:00-11:30")) // week indexing
+        assertNull(OpeningHoursTagParser.parse("2000-2044 Mo 09:00-18:30")) // year ranges
+        assertNull(OpeningHoursTagParser.parse("Mo-Fr 09:00-18:30 \"comment text\"")) // comments
+        assertNull(OpeningHoursTagParser.parse("Jan+ Mo-Fr 09:00-18:30")) //open ended month ranges
+        assertNull(OpeningHoursTagParser.parse("easter 09:00-18:00")) // easter
+        // reject multiple lists of days in a single rule, as it would change form on processing
+        // possibly with unexpected and unwanted results
+        assertNull(OpeningHoursTagParser.parse("Mo-Fr 7:30-18:00, Sa-Su 9:00-18:00"))
+    }
+
+    @Test
+    fun `accept valid opening hours`() {
+        assertNotNull(OpeningHoursTagParser.parse("Mo-Su 00:00-24:00"))
+        assertNotNull(OpeningHoursTagParser.parse("Mo 09:00-20:00"))
+        assertNotNull(OpeningHoursTagParser.parse("Tu-Su 10:00-17:30"))
+        assertNotNull(OpeningHoursTagParser.parse("We-Sa 09:30-15:00; Tu 12:00-18:30"))
+        assertNotNull(OpeningHoursTagParser.parse("Mo 9-20")) //without explicitly specified minutes
+        assertNotNull(OpeningHoursTagParser.parse("Mo-Sa 07:00-20:00; PH 7:00-7:05"))
+        assertNotNull(OpeningHoursTagParser.parse("Jun Th 17:30-19:30"))
+        assertNotNull(OpeningHoursTagParser.parse("Jul-Sep Mo 17:00-19:00"))
+        assertNotNull(OpeningHoursTagParser.parse("Jun Th 17:30-19:30; Jul-Sep Mo 17:00-19:00"))
+        assertNotNull(OpeningHoursTagParser.parse("Su 09:00-02:00")) // over midnight rules
+        assertNotNull(OpeningHoursTagParser.parse("Su 09:00-26:00")) // over midnight, +24 syntax
+        assertNotNull(OpeningHoursTagParser.parse("Mar-Oct Tu-Su 10:30-18:00; Nov-Dec Tu-Su 11:00-17:00"))
+        assertNotNull(OpeningHoursTagParser.parse("Mar-Oct Tu-Su 10:30-18:00; Mar-Oct Mo 10:30-14:00; Nov-Dec Tu-Su 11:00-17:00; Nov-Dec Mo 11:00-14:00"))
+        assertNotNull(OpeningHoursTagParser.parse("Th 17:30-19:30 open"))
+        assertNotNull(OpeningHoursTagParser.parse("Mo-Su 09:00-12:00, 13:00-14:00")) // allow time gap
+        assertNotNull(OpeningHoursTagParser.parse("Su-Mo 09:00-12:00")) // try to trigger bugs with range going end of week
+        assertNotNull(OpeningHoursTagParser.parse("Fr-Mo 09:00-12:00")) // try to trigger bugs with range going end of week
+        assertNotNull(OpeningHoursTagParser.parse("Mo 17:30-19:30; Th 17:00-19:00")) // multiple rules, no overrides
+
+        // over weekend rules, without triggering collision detector
+        assertNotNull(OpeningHoursTagParser.parse("Su-Mo 09:00-12:00; Tu-Sa 10:10-10:11"))
+        assertNotNull(OpeningHoursTagParser.parse("Fr-Mo 09:00-12:00; Tu-Th 10:10-10:11"))
```

Comment:
Maybe for all the rules that are accepted it would make sense to also check if parsing and saving again results in the same syntax?

Reply:
And parsing and saving again will check anyway whatever parsing is possible, so `assertNotNull` tests may be dropped.

Thanks to more compact test form such check is now reasonable.

Reply:
`Jun Mo 10:00-11:00;Jul Mo 11:00-12:00` is now formatted as `Jun Mo 10:00-11:00; Jul Mo 11:00-12:00` what is both more readable and preferred by https://openingh.ypid.de/evaluation_tool/

Thanks for the idea